Document alpha command usage (#762)
kyma-project/cli: docs/03-03-examples-alpha.md

---
title: Kyma CLI alpha command usage examples
type: Details
---

> The following examples show how to provision a cluster, deploy Kyma, and <!--- update is not among the alpha commands ---> with alpha commands.

Reviewer: I assume you wanted to write a comment here with `<!---` because something is still missing. But since it's inside the text instead of on a new line, it does show up on the [preview](https://github.com/kyma-project/cli/blob/2bb60b8fcf672962715c0e113132b0ea451e4a44/docs/03-03-examples-alpha.md), so you would need to delete it in the end (or write it on a separate line).

```suggestion
The following examples show how to provision a cluster, deploy Kyma, and <!--- update is not among the alpha commands ---> with alpha commands.
```

> The following examples show how to provision a cluster, deploy Kyma, and <!--- update is not among the alpha commands ---> with alpha commands.

Reviewer: I assume you wanted to write a comment here with `<!---` because something is still missing. But since it's inside the text instead of on a new line, it does show up on the [preview](https://github.com/kyma-project/cli/blob/2bb60b8fcf672962715c0e113132b0ea451e4a44/docs/03-03-examples-alpha.md), so you would need to delete it in the end (or write it on a separate line).

```suggestion
The following examples show how to provision a cluster, deploy Kyma, and <!--- update is not among the alpha commands ---> with alpha commands.
```

## Provision a cluster locally or using cloud providers
<!---provision gke is not among the alpha commands--->
To provision a cluster on a specific cloud provider (in this example GCP), run:

```bash

> kyma alpha provision gke --credentials {SERVICE_ACCOUNT_KEY_FILE_PATH} --name {CLUSTER_NAME} --project {GCP_PROJECT} 

Reviewer 2: There is only one new provision command in alpha. That is k3s. So the commands listed here will not even work..

Author: Alright, I threw out gke and minikube and added k3s instead.
Didn't find `--credentials` and `--project` among the alpha provision flags though, and none of the existing flags is an easy match. So I'm gonna assume that k3s doesn't require a service account key or a project. 

```suggestion
kyma alpha provision k3s --name {CLUSTER_NAME} 
```

```

To provision a Minikube cluster, run:
<!---provision minikube is not among the alpha commands--->
```bash
kyma alpha provision minikube
```

## Deploy Kyma

To deploy Kyma using your own domain, run:

```bash
kyma alpha deploy --domain {DOMAIN} --tls-cert {TLS_CERT} --tls-key {TLS_KEY}
```

> `{TLS_CERT}` and `{TLS_KEY}` are the TLS certificate and TLS key for the domain used. Note that both of them must be encoded in base64. You can use [OpenSSL](https://www.openssl.org/) (which is installed by default with macOS) to convert your TLS certificate and TLS key to base64-encoded values.

Reviewer 2: The arguments changed. The cert and key must be passed as a file now. Please provide a full example with a self-signed certificate and a fake domain, then people can try it out

Reviewer 3: @NHingerl : users can create a self-signed certificate by running this command:
`openssl req -x509 -newkey rsa:4096 -keyout key.pem -out crt.pem -days 365`

Author: I assume creating the self-signed cert is a prerequisite to be done before running the `alpha deploy` command, right?

I'm not sure what to put in a full example. Pls provide sample text.

> You can simplify the process to deploy Kyma with your own domainin  `{DOMAIN}` the following way:

Reviewer 3: ```suggestion
You can simplify the process to deploy Kyma with your own domain `{DOMAIN}` in the following way:
```

1. Encode the TLS certificate and TLS key, and export them as environment variables:
  
    ```bash
    export TLS_CERT="$(openssl base64 -in {TLS_CERT_FILE_PATH})"
    export TLS_KEY="$(openssl base64 -in {TLS_KEY_FILE_PATH})"
    ```

    `{TLS_CERT_FILE_PATH}` is the path to the file containing the TLS certificate, and `{TLS_KEY_FILE_PATH}` is the path to the file containing the TLS key.

2. Use these environment variables in the command:

    ```bash
    kyma alpha deploy --domain {DOMAIN} --tls-cert $TLS_CERT --tls-key $TLS_KEY
    ```

> 

Reviewer 3: You can merge step 1 and 2 to:

1. Create a self-signed TLS certificate files:
`openssl req -x509 -newkey rsa:4096 -keyout key.pem -out crt.pem -days 365`
2. Pass the certificate files to the deploy command
`kyma alpha deploy --domain {DOMAIN} --tls-cert crt.pem --tls-key key.pem`

Author: is it `--tls-cert` (as in this example) or `--tls-crt` (as in the command docu) ?

To deploy Kyma from the `master` branch, run:

```bash

> kyma alpha deploy --source master

Reviewer 3: Not sure if it will be too redundant, but beside `master` and  `local` are also a few more options supported (e.g. pull-request or revision etc.)

Author: Good point. I think I'll mostly refer to the command docu for this to avoid redundancy and thus the maintenance effort.

```

To deploy Kyma using your own Kyma installer image, run:

```bash

> kyma alpha deploy --source {IMAGE}

Reviewer 2: That usage is not existing anymore, concept of installer image is gone

Author: Removing that section.

```

To build an image from your local sources and deploy Kyma on a remote cluster based on this image, run:

```bash

> kyma alpha deploy --source local --custom-image {IMAGE}

Reviewer 2: Picking as source the local checkout is still supported, but you cannot specify a custom image anymore as it is simply not required and possible anymore, we simplified a lot here

Reviewer 3: When using the `local` source option, we should point out that the CLI will per default expect the sources in the `./workspace` folder  (which is probably not the location where the user stores the Kyma sources locally). He can adjust the workspace-path by setting the `-w ${path_to_workspace}` parameter.

Author: Alright! This is what I got now:

To build Kyma from your local sources and deploy it on a remote cluster, run:

```bash
kyma alpha deploy --source=local
```
> **NOTE:** By default, Kyma expects to find the sources in the `./workspace folder`. To adjust the path, set the `-w ${PATH_TO_WORKSPACE}` parameter.

```

> >    **EXAMPLE:** If your `{IMAGE}` is `user/my-kyma-installer:v1.4.0`, this command builds an image from your local sources and push it your repository `user/my-kyma-installer` with the tag `v1.4.0`.<br>

Reviewer: We normally use quote blocks only for note, tip, and caution panels, see https://kyma-project.io/community/guidelines/content/#formatting-formatting-panels
Also, I think you can either write "will build... push" or "builds... pushes" to have the right verb agreement. 
```suggestion
For example, if your `{IMAGE}` is `user/my-kyma-installer:v1.4.0`, this command builds an image from your local sources and pushes it to your repository `user/my-kyma-installer` with the tag `v1.4.0`.<br>
```

Author: Alright, noted! I think this particular section is obsolete anyway, but I'll keep it in mind. Thank you for noticing the typo :D

Then, this image `user/my-kyma-installer:v1.4.0` is used in the `Deployment` for `kyma-installer`.

> To deploy kyma with only specific components, run:

Reviewer: ```suggestion
To deploy Kyma with only specific components, run:
```

Author: Thank you! corrected.

```bash
kyma alpha deploy --components {COMPONENTS_FILE_PATH}
```
`{COMPONENTS_FILE_PATH}` is the path to a YAML file containing the desired component list to be installed, for example:
```bash
  components:
  - name: "cluster-essentials"

>     namespace: "kyma-system"

Reviewer 2: Please remove the neamesace entries having "kyma-system" as this namespace will be used by default, no need to specify it

Reviewer 3: Here an example of the latest components file: https://github.com/kyma-project/kyma/blob/master/installation/resources/components.yaml

Maybe we need a short explanation how the structure of this file has to be (@a-thaler : maybe worth to think about a schema validation in the CLI which detects invalid formats).

Author: What I got now is:

o deploy Kyma with only specific components, run:

```bash
kyma alpha deploy --components {COMPONENTS_FILE_PATH}
```
`{COMPONENTS_FILE_PATH}` is the path to a YAML file containing the desired component list to be installed, for example:
```bash
  components:
  - name: "cluster-essentials"
  - name: "testing"
  - name: "istio"
    namespace: "istio-system"
  - name: "xip-patch"
    namespace: "kyma-installer"
  - name: "istio-kyma-patch"
    namespace: "istio-system"
  - name: "dex"
  ```
  
I'll need input for the explanation that Tobias suggested.

  - name: "testing"
    namespace: "kyma-system"
  - name: "istio"
    namespace: "istio-system"
  - name: "xip-patch"
    namespace: "kyma-installer"
  - name: "istio-kyma-patch"
    namespace: "istio-system"
  - name: "dex"
    namespace: "kyma-system"
  ```
In this example, only these 6 components are deployed on the cluster.

To override the standard Kyma installation, run:
<!---didn't find override among the deploy options--->
```bash

> kyma alpha deploy --override {OVERRIDE_FILE_PATH}

Reviewer 2: That changed to the --values approach, overrides is the old concept. Please adjust to it. Remember that we now support providing values by file and by arguments, see the inline help :)

Reviewer 3: We have not two options to provide overrides:

1. As file by setting `--values-file file1.yaml --values-file file2.yaml` or in short-form as`-f file1.yaml...`
1. As single value by setting `--value comp.key=value --value comp.key2=value` (here we don't offer a short-form)

Author: That's good to know! I was wondering there the --values come in. 
Do we still use the word "override" in prose text? As in:

To override the standard Kyma installation, you can use the `--values` flag in the following ways:
- Override a file by setting `--values-file file1.yaml --values-file file2.yaml` (short form: `-f file1.yaml...`)
- Override a single value by setting `--value comp.key=value --value comp.key2=value` (no short form available)

```
`{OVERRIDE_FILE_PATH}` is the path to a YAML file containing the parameters to override, for example:
```bash
  apiVersion: v1
  kind: ConfigMap
  metadata:
    name: ory-overrides
    namespace: kyma-installer
    labels:
      installer: overrides
      component: ory
      kyma-project.io/installation: ""
  data:
    hydra.deployment.resources.limits.cpu: "153m"

>     hydra.deployment.resources.requests.cpu: "53m"

Reviewer 3: The new override file format looks like:

    componentName:
      overrideKey: overrideValue
      overrideKey2:
        overrideSubKey: overrideValue2

  ---
  apiVersion: v1
  kind: ConfigMap
  metadata:
    name: monitoring-overrides
    namespace: kyma-installer
    labels:
      installer: overrides
      component: monitoring
      kyma-project.io/installation: ""
  data:
    alertmanager.alertmanagerSpec.resources.limits.memory: "304Mi"
    alertmanager.alertmanagerSpec.resources.requests.memory: "204Mi"
  ```
In this example, overrides are provided for 2 different components: `ory` and `monitoring`.
 
- For `ory`, the values of `hydra.deployment.resources.limits.cpu` and `hydra.deployment.resources.requests.cpu` will be overriden to `153m` and `53m` respectively.
- For `monitoring`, the values of `alertmanager.alertmanagerSpec.resources.limits.memory` and `alertmanager.alertmanagerSpec.resources.requests.memory` will be overriden to `304Mi` and `204Mi` respectively.

> It is also possible to provide multiple override files at the same time

Reviewer: ```suggestion
It is also possible to provide multiple override files at the same time:
```

Author: Thanks!

  ```bash
  kyma install --override {OVERRIDEL_FILE_1_PATH} --override {OVERRIDE_FILE_2_PATH}
  ```

## Upgrade Kyma
<!---upgrade is not among the alpha commands--->

> 

Reviewer 3: The upgrade will also do a compatibility check between the new and the currently installed version. We should mention, that this version check can only deliver reliable feedback if a Kyma release was installed (not from a PR, branch, revision or local).

Author: That's good to know! Added it.

> To upgrade the Kyma version on the cluster, you can run the `upgrade` command which has the same structure and same flags as the `install` command.

Reviewer 2: Upgrade and install are now covered both by "deploy", so please assure that you cover upgrade scenarios in the above section.

Author: I'd prefer to keep a< separate headline because that's probably where users might look for it; but then refer to the "Deploy" section for more details.

> ## Test Kyma

Reviewer 2: Testing is based on existing stable commands, not worth to mention at all in the alpha section

Author: Excellent. Dropped the whole section.

<!---test is not among the alpha commands--->

To check which test definitions are deployed on the cluster, run:

```bash
kyma test definitions
```

To run all the tests, run:

```bash
kyma test run
```

To check the test results, run:

```bash
kyma test status
```

> ## Use the *--values* flag

Reviewer: ```suggestion
## Use the --values flag
```

Author: Testing section is thrown out. Thanks though!

> The *--values* flag can be helpful for the following cases: 

Reviewer: For flags, we normally use the code font, see https://kyma-project.io/community/guidelines/content/#formatting-formatting-use-code-font-for-these-items

```suggestion
The `--values` flag can be helpful for the following cases: 
```

Author: got it!

> - To set the administrator passwort, use:

Reviewer: ```suggestion
- To set the administrator password, use:
```

Author: LOL thanks for noticing :D
